nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/PasswordBasedOauth2TokenProvider.java

```diff
@@ -0,0 +1,230 @@
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.annotation.lifecycle.OnEnabled;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.components.Validator;
+import org.apache.nifi.controller.AbstractControllerService;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.expression.ExpressionLanguageScope;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.ssl.SSLContextService;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.X509TrustManager;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+@Tags({"oauth2", "provider", "authorization", "access token", "http" })
+@CapabilityDescription("Provides OAuth 2.0 access tokens that can be used as Bearer authorization header in HTTP requests." +
+    " Uses Resource Owner Password Credentials Grant.")
+public class PasswordBasedOauth2TokenProvider extends AbstractControllerService implements OAuth2AccessTokenProvider {
+    public static final PropertyDescriptor AUTHORIZATION_SERVER_URL = new PropertyDescriptor.Builder()
+        .name("authorization-server-url")
+        .displayName("Authorization Server URL")
+        .description("The URL of the authorization server that issues access tokens.")
+        .required(true)
+        .addValidator(StandardValidators.URL_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .build();
+
+    public static final PropertyDescriptor USERNAME = new PropertyDescriptor.Builder()
+        .name("service-user-name")
+        .displayName("Username")
+        .description("Username on the service that is being accessed.")
+        .required(true)
+        .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .build();
+
+    public static final PropertyDescriptor PASSWORD = new PropertyDescriptor.Builder()
+        .name("service-password")
+        .displayName("Password")
+        .description("Password for the username on the service that is being accessed.")
+        .required(true)
+        .sensitive(true)
+        .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+        .build();
+
+    public static final PropertyDescriptor CLIENT_ID = new PropertyDescriptor.Builder()
+        .name("client-id")
+        .displayName("Client ID")
+        .required(false)
+        .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .build();
+
+    public static final PropertyDescriptor CLIENT_SECRET = new PropertyDescriptor.Builder()
+        .name("client-secret")
+        .displayName("Client secret")
+        .dependsOn(CLIENT_ID)
+        .required(true)
```

Comment:
This property is required, but it appears to be ignored if `Client ID` is not specified.

Reply:
Yes that's exactly what I wanted to achieve. It depends on CLIENT_ID. If that is set, this should be set as well. If CLIENT_ID is not set, this can be ignored.

Reply:
Thanks for clarifying, I missed the `dependsOn()` attribute in the initial review.